src/common/ceph_clock.h

```diff
@@ -0,0 +1,420 @@
+      std::is_same<T, coarse_real_time>::value ||
+      std::is_same<T, mono_time>::value ||
+      std::is_same<T, coarse_mono_time>::value>* = nullptr>
+  auto now() -> T {
+    return get_time_point_<T>();
+  }
+
+  template <typename T,
+    typename std::enable_if_t<
+      std::is_same<T, real_time>::value ||
+      std::is_same<T, coarse_real_time>::value ||
+      std::is_same<T, mono_time>::value ||
+      std::is_same<T, coarse_mono_time>::value>* = nullptr>
+  void reset() {
+    get_time_point_<T>();
+  }
+
+  template <typename T,
+    typename std::enable_if_t<
+      std::is_same<T, real_time>::value ||
+      std::is_same<T, coarse_real_time>::value ||
+      std::is_same<T, mono_time>::value ||
+      std::is_same<T, coarse_mono_time>::value>* = nullptr>
+  auto elapsed_from_start() -> ceph::timespan {
+    return elapsed_from_start_<T>();
+  }
+
+  template <typename T,
+    typename std::enable_if_t<
+      std::is_same<T, real_time>::value ||
+      std::is_same<T, coarse_real_time>::value ||
+      std::is_same<T, mono_time>::value ||
+      std::is_same<T, coarse_mono_time>::value>* = nullptr>
+  auto elapsed_from_now() -> ceph::timespan {
+    return elapsed_from_now_<T>();
+  }
+
+  template <typename T, typename S,
+    typename std::enable_if_t<
+      std::is_same<T, nanoseconds>::value ||
+      std::is_same<T, milliseconds>::value ||
+      std::is_same<T, seconds>::value &&
+      std::is_same<S, coarse_real_time>::value ||
+      std::is_same<S, coarse_mono_time>::value>* = nullptr>
+  void set_duration(ceph::mono_clock::duration& d, int64_t s) {
+    set_duration_<T>(d, s);
+  }
+
+  template <typename T, typename S,
+    typename std::enable_if_t<
+      std::is_same<T, nanoseconds>::value ||
+      std::is_same<T, milliseconds>::value ||
+      std::is_same<T, seconds>::value &&
+      std::is_same<S, coarse_real_time>::value ||
+      std::is_same<S, coarse_mono_time>::value>* = nullptr>
+  double get_count() {
+    return get_count_<T, S>();
+  }
+
+  /* Compatibility: utime_t */
+  template <typename T,
+    typename std::enable_if_t<
+      std::is_same<T, utime_t>::value>* = nullptr>
+  T now() {
+    ut_now_ = ceph_clock_now_();
+    return ut_now_;
+  }
+
+  template <typename T,
+    typename std::enable_if_t<
+      std::is_same<T, utime_t>::value>* = nullptr>
+  T elapsed_from_start() {
+    ut_end_ = ceph_clock_now_();
+    return ut_end_ - ut_start_; 
+  }
+
+  template <typename T,
+    typename std::enable_if_t<
+      std::is_same<T, utime_t>::value>* = nullptr>
+  T elapsed_from_now() {
+    ut_end_ = ceph_clock_now_();
+    return ut_end_ - ut_start_;
+  }
+
+  template <typename T,
+    typename std::enable_if_t<
+      std::is_same<T, utime_t>::value>* = nullptr>
+  void reset() {
+    ut_now_ = ceph_clock_now_();
+  }
+
+  ostream& set_localtime(ostream& out, utime_t& ut) const {
+    return ut.localtime(out);
+  }
+
+private:
+  /** real_clock **/
+  real_time rt_start_;
```

Comment:
@shinobu-x why do these private member variables end in underscores?  I haven't seen that convention.

Reply:
@mattbenjamin i've always added `_` to distinguish public members.

Reply:
There doesn't seem to be a great need to distinguish something very clear to the compiler.

Reply:
@mattbenjamin i don't mind deleting because adding `_` is not quite the purpose of this PR if you want.